### ISSUE TITLE
Add a module system

### ISSUE THREAD

**Issue by JulienCabanes** [sass/sass#1094]
- [x] [Proposal](https://github.com/sass/language/blob/master/accepted/module-system.md)
- [x] [Specs](https://github.com/sass/sass-spec/tree/feature.use)
- [x] [Documentation](https://github.com/sass/sass-site/issues/263)
- [x] [Dart Sass](https://github.com/sass/dart-sass/issues/554)
- [ ] [LibSass](https://github.com/sass/libsass/issues/2807)

See also [the module system project](https://github.com/orgs/sass/projects/3).

---

This is a compilation of issues and wishes about @import
### Import CSS files #556

``` scss
@import "normalize";
```

Would import sass, scss or **css** file

> There's a [plugin](https://github.com/chriseppstein/sass-css-importer) for that
### Import Once #139

``` scss
@import-once "foo";
@import "foo" !once;
@depend "foo";
```

Multiple syntax suggestions for one time import

> There is a [plugin](https://github.com/chriseppstein/compass/tree/master/import-once) for that
### Import Directory #690

`@import "foo";`

If "foo" is a folder than it tries to import "foo/_module.scss"

> reasonable
### Namespaces and Aliases #353

``` scss
@import "foo" as "bar";
```

This would put every "foo" components (mixins, vars, placeholders...) in a "bar" package.

What would be the separator ? `.` or `/` or `\` or `::` ?
Where would go the `$` and `%` ? Before or after the namespace ?
What it would look like :

``` scss
// Mixins
@include bar.baz();
or
@include bar/baz();

// Variables
color: $bar.baz;
or
color: bar.$baz;
or
color: $bar/baz;
or
color: bar/$baz;

// Placeholders
div { @extend %bar.baz; }
or
div { @extend bar.%baz; }
or
div { @extend %bar/baz; }
or
div { @extend bar/%baz; }

// Functions
color: bar.baz();
or
color: bar/baz();
```
### Non-transitive imports [#353](https://github.com/nex3/sass/issues/353#issuecomment-5146513)

> if A imports B and B imports C, A shouldn't necessarily see everything from C.
> A way of getting around non-transitive imports.
### Renaming classes [#353](https://github.com/nex3/sass/issues/353#issuecomment-5146513)

> This isn't something I'm 100% sold on, but it has been requested. Maybe the correct way to do this is some combination of namespacing and @extend.
### Disabling CSS output [#353](https://github.com/nex3/sass/issues/353#issuecomment-5146513)

> Both selectively and wholesale. The importer may want to @extend the rules in the importee but not have them concretely present, or it may want to suppress rules matching a certain selector.

This is also related to #320 :

``` scss
// A.scss
.foo {...}
.bar {...}

// B.scss
@use "A";
.baz { @extend .foo; }
```

B output would not contain `.foo` & `.bar` definition but only `.baz`;
### Conditional imports #451

``` scss
@if true {
    @import "foo";
}
```

It "worked" before but never intended.
### Paths

A way to setup shortcut paths (like [Require.js](https://github.com/jrburke/r.js/blob/master/build/example.build.js#L46-L49)) 

``` scss
// No syntax intended, only the purpose
@paths (
    "bootstrap": "../components/bootstrap-sass/vendor/assets/stylesheets/bootstrap"
);

@import "bootstrap";
```

**Comment by lolmaus** [comment]
Wow, great job @JulienCabanes.

**Comment by Anahkiasen** [comment]
In a way there is a plugin for [import directory](https://github.com/chriseppstein/sass-globbing) but since it's from Chris I'm pretty sure they know about it :p

**Comment by chriseppstein** [comment]
Adding #739 to the list which is necessary for #451 as well as any use of SassScript in an import directive.

**Comment by KittyGiraudel** [comment]
> If "foo" is a folder than it tries to import "foo/_module.scss"

I really don't understand where do you get this from. I'd expect folder include to behave like this:

``` scss
// Import a single file
@import "file"; 
// Import all files from folder
@import "folder/*"; 
// Import all files from all folders from folder
@import "folder/**/*";
```

Or am I wrong?

**Comment by JulienCabanes** [comment]
@HugoGiraudel this came from #690 
The problem with `folder/*` is you can't control the import order, you would expect an alphabetical one but it's too implicit.
Having a standard `_module.scss` responsible for importing in the correct order would be better.

**Comment by Anahkiasen** [comment]
Why not both ? Use an index file if present, otherwise alphabetical ?

**Comment by jslegers** [comment]
@JulienCabanes & @HugoGiraudel & @Anahkiasen :

I love the suggestion made by @Anahkiasen.

An alternate solution would be to have a different syntax depending on what you're loading:
- If the path ends with a letter of number, it's considered a file
- If the path ends with `/`, it's considered a folder and an index file will be loaded... which could be named `_module.scss`, `_index.scss`, `_component.scss`, `_loader.scss`, ... Maybe the devs could leave it do users to configure what name they want for their index file? I'm pretty sure we're all unlikeoy to want to give up our personal preference here.
- If the path ends with `/*`, it's considered a folder and all files will be loaded alphabetically.

## Example code

``` scss
// Import a single file named "file"
@import "file"; 
// Import file "_index.scss" from folder "base"
@import "base/"; 
// Import all files from folder "base" in alphabetical order
@import "base/*"; 
// Import file "_index.scss" from every subfolder of folder "base",
// with folders loaded in alphabetical order
@import "base/*/";
// Import all files from every subfolder of folder "base"
// with both folders and files loaded in alphabetical order
@import "base/*/*";
```

**Comment by KittyGiraudel** [comment]
Convention is to use `**` for folders, `*` for files. Henre `@import "**/*"`.

**Comment by jslegers** [comment]
@HugoGiraudel :

I'm not familiar with that convention, but it's but a minor difference...

``` scss
// Import a single file named "file"
@import "file"; 
// Import file "_index.scss" from folder "base"
@import "base/"; 
// Import all files from folder "base" in alphabetical order
@import "base/*"; 
// Import file "_index.scss" from every subfolder of folder "base",
// with folders loaded in alphabetical order
@import "base/**/";
// Import all files from every subfolder of folder "base"
// with both folders and files loaded in alphabetical order
@import "base/**/*";
```

**Comment by robwierzbowski** [comment]
"Renaming classes #353" could be accomplished with "Disabling CSS output #353". Disabling selector output could be said another way — "transform selector to placeholder".

```
// Pseudocode, no syntax implied
.class {
  color: red;
}

@no-output .class;

.module {
  @extend .class;
}
```

**Comment by apfelbox** [comment]
@Anahkiasen I would just do the alphabetical import, without the automatic check for the `_modules.scss` file. Because if I care for the order of my imports (and can't name them `_1-modules.scss` or something like that = I already have to be specific) I can be so specific to just write `@import "foo/modules";`

**Comment by aripalo** [comment]
This sort of falls under the conditional imports #451, but slighlty different use case: 

Currently trying to declare `@import` within mixin will result to `Error: Import directives may not be used within control directives or mixins.`-Error.

What I am trying to do is following:

```
@mixin component($name) {
  .#{$name} {
    @import '/components/#{$name}/#{$name}';
  }
}

@include component('foo');
```

This would be a great way to enforce "namespaced" styles for the components (in my case, Angular directives which live in their own separate folders with their javascripts, templates, tests **and styles**), and also allow nice short syntax (that `@include component('foo');`) for importing them.

**Comment by koenpunt** [comment]
@robwierzbowski a css placeholder can be defined with `%`

``` scss
%class-name {
  color: red;
}

.module {
  @extend %class-name;
}
```

**Comment by andi1984** [comment]
I hav to agree with @aripalo:

Some sort of `@import` directive inside a mixin would be great, especially for including third party styles.

**Comment by chriseppstein** [comment]
@andi1984 see #739

**Comment by RaphaelDDL** [comment]
+1 for Paths via variables but _specially_ via something @require-like. 

Having aliases for css module partials would be awesome. Implementing a `_sass_config.scss` to be imported on all files like we do with `require_config.js` on html _(or even having a default file name at project sass root folder that would be self-imported, following the idea of `@import 'folder'` importing a default `_modules.scss` inside that folder)_.

An addition to the idea is to have a keyword pointing to project root folder, for example:

```
@path(
    'myProjRoot': :sass_root
);
```

This way we leave sass to resolve the root folder path (as some projects are run in multiple OSes so Windows is C:/ but Mac ~/ and so on) and also because sometimes we want to import one single file - without making a full path alias for it - and having `@import "myProjRoot/features/aThing/something"` is so much easier/faster to understand than `@import "../../../../../../../../../../../features/aThing/something"`.

**Comment by backnight** [comment]
Hello, I would like to know where is the current development state of these features ? Especially for a feature like this in less : http://lesscss.org/features/#import-options

=> @import (reference) "filename";

Thank you.

**Comment by Jakobud** [comment]
The ability to use a variable name for your @import string would be very nice. Right now this functionality is missing from SASS.

**Comment by sammorrisdesign-zz** [comment]
I'll +1 the ability to use `@import` inside of a mixin with variable names. We just tried to write something like this for our third party components.

``` sass
 @mixin styles-for($list) {
     @each $component in $list {
         @import '#{$component}/_styles.scss';
     }
 }
```

which would be called with...

``` sass
@include styles-for(grid typography buttons);
```

Which would allow us to add new components with less maintenance.

**Comment by Jakobud** [comment]
Wait @sammorrisdesign does that mixin actually work? How are you able to import using a variable $component?

**Comment by sammorrisdesign-zz** [comment]
@Jakobud Nah, we tried to write something like that. But it didn't work. Hence the +1 to be able to use `@import` inside mixins and with variable names.

Would be great if it did though ;)

**Comment by jlukic** [comment]
Would kill for #739, so we can finish converting Semantic UI to SASS which our users would really love. We even mention it [in our readme](https://github.com/Semantic-Org/Semantic-UI#places-to-help).

**Comment by nex3** [comment]
@jlukic Why?

**Comment by Jakobud** [comment]
@nex3 are you asking why someone would want variables in @import strings? Simple:

Lets say you have a file that imports a couple dozen other files:

```
@import "my/path/file1";
@import "my/path/file2";
@import "my/path/file3";
@import "my/path/file4";
@import "my/path/file5";
etc...
```

Then lets say you have to change `my/path/` to something else. You now how to change it for all the import strings. If you could use variables though you could have something like this:

```
$myPath = "whatever/path/i/want/";

@import $myPath + "file1";
@import $myPath + "file2";
@import $myPath + "file3";
@import $myPath + "file4";
@import $myPath + "file5";
etc...
```

It makes changing tons of import strings a piece of cake and opens the door for dynamic build systems.

**Comment by cimmanon** [comment]
@Jakobud Read the linked issue again, that's not what's being asked for.  You're asking for a completely different issue that can be solved by features that already exist in Sass (see: https://github.com/sass/sass/issues/279#issuecomment-100204040)

**Comment by sulfo** [comment]
+1 for a LESS like feature of `import (reference) "filename";` as @backnight asked for in his comment. It's the only thing there's holding me back from switching to SASS.

**Comment by s2software** [comment]
+1 for LESS like `import (reference) "filename"` too.

**Comment by dehuszar** [comment]
Not to turn this into a +1 thread, but I also cannot use SASS without `@import (reference) 'filename'`.  It is the number one way of allowing reuse of css objects/patterns while keeping compiled output sizes to a minimum.

**Comment by Jakobud** [comment]
@cimmanon 

> You're asking for a completely different issue that can be solved by features that already exist in Sass (see: #279 (comment))

That only applies for someone using vanilla Ruby SASS at the command line. Not easy (or not possible in many cases) for everyone using SASS via CodeKit, Grunt, Gulp, ANT, Node, and a 100 other build tools with custom SASS pipelines.

**Comment by yuki-takei** [comment]
+1 for a LESS like feature of import (reference) "filename".

**Comment by kouki-o** [comment]
+1 @import (reference) "filename";

**Comment by janaspage** [comment]
+1 for `@import` that can handle variables, and/or `@import` that can be used within `@if` directives, and/or `@import` that can handle values returned by `@function` directives.

**Comment by p2media** [comment]
+1 for Import Once, Import Directory, Namespaces and Aliases and Paths :)
Especially the shortcuts.

**Comment by chriseppstein** [comment]
There's some good stuff in here worth reading and considering: https://github.com/css-modules/css-modules

**Comment by meishier** [comment]
+1 @import (reference) "filename";

**Comment by nex3** [comment]
@chriseppstein That's going in a cool direction, but the README is super vague.

**Comment by Lewiscowles1986** [comment]
:+1: if you could @import (reference) "filename"; Chris Coyier just highlighted my ignorance of this, and now I will LESS my SASS output when done to see what that makes of it.

**Comment by backnight** [comment]
Just released yesterday => https://css-tricks.com/reference-imports-in-less-are-kinda-cool/

Enjoy guys !

**Comment by jfurrow** [comment]
+1 for `@import (reference) ...`!

**Comment by gherkins** [comment]
+1

> +1 for @import (reference) ...!

**Comment by JulienCabanes** [comment]
As suggested by @chriseppstein, cssmodules has cool stuffs like [composition and dependencies](https://github.com/css-modules/css-modules#dependencies) which looks like LESS's `@import (reference)`

``` css
.otherClassName {
  composes: className from "./style.css";
}
```

**Comment by xzyfer** [comment]
Personally I think there is something promising in the approach [ICSS](https://github.com/css-modules/icss) takes. I think something like this gives us a good foundation for which to do namespacing/aliasing and paths, as well as better approach to class renaming (by solving the root problem).

**Comment by xzyfer** [comment]
I think there is some great work happening in this space that we can leverage. For example - http://glenmaddern.com/articles/interoperable-css

**Comment by davidkpiano** [comment]
@xzyfer Here's some of my initial thoughts on how ICSS imports, aliasing, and other constructs can look like in SCSS and using W3 grammar on [media features](http://www.w3.org/TR/mediaqueries-4/#mq-features), as well as using `@import` as an [at-rule that accepts a block](http://www.w3.org/TR/css-syntax-3/#at-rule):

``` scss
// _partial.scss

// The @export at-rule can contain variables, rulesets, mixins, etc.
@export {
  $foo-color: blue;
  $bar-color: red;

  %alias { @extend .baz; }
}

// Selectors are always exported as-is
.foo {
  color: $foo-color;
}

.bar > .baz {
  color: $bar-color;
}
```

**Using @import with (reference) media feature**

``` scss
// main.scss

// Uses "reference" as a named media feature that does not match any device in CSS (which is fine).
// Treats every selector as a silent selector (similar to a %placeholder)
@import '_partial.scss' (reference);

.test {
  @extend .foo; // extends .foo from _partial.scss, but .foo is never output
}

.test2 {
  @extend .baz; // extends .baz from _partial.scss, resulting in .bar > .test2
}
```

**Using `@import` as an at-rule with block:**

``` scss
// _another-partial.scss (using @import as an at-rule with block)

// $foo-color and $bar-color are NOT visible outside this at-rule,
// unless they're explicitly defined inside this _another-partial.scss file.
@import '_partial.scss' (reference) {
  $color-primary: $foo-color;
  $color-secondary: $bar-color;

  // Selectors can also be aliased safely here, using @extend like normal:
  .my-foo { @extend .foo; }
}

.my-foo {
  display: block;
  background-color: $color-secondary;
}

// Results in:
// .my-foo {
//   color: blue;
//   display: block;
//   background-color: red;
// }

.aliased-baz {
  @extend %alias; // from _partial.scss
}
```

**Overloading `@extend`**

``` scss
@import '_partial' (reference);
@import '_partial2' (reference); // suppose this also contains .foo

.foo-from-first {
  @extend .foo from '_partial';
}

.foo-from-second {
  @extend .foo from '_partial2';
}
```

In summary, I'm not a huge fan of how ICSS uses the syntax for properties and values outside of an at-rule to define imports and exports, and am _especially_ against the use of pseudo-selectors (that don't select anything) where at-rules are more appropriate.

We can make minimal changes to SCSS syntax and still get all the benefits of how imports, exports, local aliases, etc. are done in ICSS, with greater adherence to W3C syntax grammar. The changes are:
- Giving Sass comprehension of certain media features, such as `(reference)`, with regard to `@import`
- Allowing `@import` to contain a block
- Overloading `@extend` to comprehend `from` as such:
  - `@extend <string> from [ <url> | <string> ];`

What do you think?

**Comment by nex3** [comment]
It's interesting to look at ICSS for inspiration, but we're not going to be directly borrowing the syntax. We're also moving away from the `@import` keyword in general, so there's no need to be tethered to valid CSS syntax (not that there really was anyway).

**Comment by long-lazuli** [comment]
For consistency, I would prefer `@import '_partial' !reference;`

``` SCSS
// What about
@import '_big-library' !reference as lib;
// or
@import '_small-library' !reference(small);
button {
   @extend (lib) .big-fat-button;
   @extend (small) .button__typo;
}
```

**Comment by chriseppstein** [comment]
@long-lazuli we're introducing a new directive and @import will be purely for generating a CSS import (after a deprecation period). So consistency is not an overriding concern. Good syntax is the main goal.

**Comment by long-lazuli** [comment]
@chriseppstein Whoops, not aware of that; could you point me over this new directive's discussion ?

By the way; you have my syntax proposal.
the way we `@extend` doesn't really matter for me, but it would be usefull to name the import locally, as it would be a bit painful to write :

``` SCSS
@extend .foo from 'lib/partial/stuff/far/far/faraway';
// or
@extend ('lib/partial/stuff/far/far/faraway') .foo;
```

**Comment by yoannisj** [comment]
I have already posted this in [issue #279 ](https://github.com/sass/sass/issues/279), but that is an old issue and as wishes and suggestions for importing sass files has moved here, I repeat it here.

It would be very useful to allow aliasing in paths to import/load. Aliases could be a configuration option for the Sass Compiler. For example, we could add a [:load_aliases] option which would be similarl to the [`:load_paths` option](http://sass-lang.com/documentation/file.SASS_REFERENCE.html#load_paths-option), but offer more flexibility by taking an object where the keys can be used to point to a specific load path in `@import`.

This extra flexibility covers a lot of different use cases:
- import stylesheets directly from bower-components (see note below)
- allow for a more complex project structure, where sass files are scattered in different directories/bundles but need to compile into a single file
- swap source directories based on a config file (e.g. for theming)

**options**

``` ruby
load_aliases:
   vendorcomp: './bower_components/vendor-component/path/to/stylesheets/sass/'
   foo: './src/path/to/bundles/foo-bundle/stylesheets/',
   theme:  './src/path/to/themes/blue-orchid/stylesheets/'
```

**sass file** (can be anywhere, for example in './src/main/stylesheets/' or whatever)

``` scss
@import "vendorcomp/pagination";
@import "foo/modules/navbar";
@import "theme/colors";
```

**Comment by chriseppstein** [comment]
It's fairly straightforward to build a custom importer that implements the load aliases concept. The convention in Sass has long been to namespace within your sass distribution so that everyone imports the plugin the same way off the load path. This solution present a problem for shared transitive dependencies because there's no canonical way for all of the extensions to import the same sass file from the shared dependency since they won't know what import alias the main application has given it.

**Comment by amster** [comment]
Hi, I thought I just saw @hcatlin present on the new `@use` to pull a file in such that

```
@use "somefile";

.someclass {
  color: $somefile.somevariable;
}
```

I guess I find it it weird that `"somefile"` would turn into a variable of the same name. What if the filename has characters in it that can't turn into a variable name? Might it be clearer to extend the `@import` syntax to explicitly tell the reader what variable to use?

```
@import "somefile" as $somefile;
.someclass { color: $somefile.somevariable; }

// or

@import "somefile" as $globals-in-a-file;
.someclass { color: $globals-in-a-file.somevariable; }
```

**Comment by astroanu** [comment]
Still no css import ? :(

**Comment by HamptonMakes** [comment]
@amster I should have had that in my example, definitely `@use "file" as identifier` is something that we'll do. I think it's mentioned somewhere above. I was just showing a shorthand version that I like to have as an option.

**Comment by nex3** [comment]
To clarify, the syntax and semantics for `@use` are still very much up in the air. @hcatlin was presenting an idea that's in the vein of what we're looking at, but it's not at all final or binding.

**Comment by saas786** [comment]
**+1**
For allowing import as reference only.
`@import (reference) "somefile";`

**+1**
For allowing variable names in import
`@import "#{$rootpath}/somepath/#{$somefile}";`

**+1**
For allowing dynamic variables
Ref: https://github.com/gruntjs/grunt-contrib-less#modifyvars

**Comment by Friend-LGA** [comment]
+1 for LESS like `import (reference) "filename"`

**Comment by jlukic** [comment]
Life's tough...

**Comment by saas786** [comment]
I would also like something similar for variables:

**+1**

```
$base-color: #cf4944 !default;
```

and inside some other file which imports after "variables" file

I wish I could do something like this: 

```
$base-color: #fff !override;
```

so by default it works how its already implemented in sass,

but if I want to override some variable after its been already defined, I still could do it.

**Comment by koenpunt** [comment]
Overriding variables after definition is already supported with `!default`. 

E.g.;

``` scss
// _vars.scss
$color: blue !default;
```

``` scss
// main.scss
@import "vars";
$color: green;

body {
  background: $color;
}
```

results in 

``` css
body {
  background: green;
}
```

**Comment by saas786** [comment]
Thanks @koenpunt 

But I already tried that, and it didn't worked for me.

Using:
"grunt-sass": "^1.1.0",

Although it works in online compiler http://sassmeister.com/

Any clue, why it would not work for me?

**Comment by saas786** [comment]
I did give it another try,

this time I did write some simple scss code, just to verify whether my grunt setup is incorrect or I have some other issue, and for simple code it worked.

Then I did some digging into my code, and found out that if I try override variables after the code file has already been imported (where this variable is going to use), then it will not work.

Note: Ideally it should have worked, as compilers first import all variables / code and only then compile them. Not sure why sass compiler didn't worked for me.

Like

`// _vars.scss
$color: blue !default;`

`// main.scss
body {
  background: $color;
}`

`// _override_vars.scss
$color: green;`

which I was doing in (LESS), and tried to keep same structure in SASS.

Thanks @koenpunt for pinging me about it, and it forced to me give it another try, otherwise I had given up on it, and rewritten my code already, but I like this approach :)

**Comment by koenpunt** [comment]
I think the way this works (can override before usage, and not after) is good, because this way you can use variable names across your project, without inadvertently altering previous usage.

**Comment by jslegers** [comment]
@koenpunt 

What exactly is the difference between...

``` scss
$color: blue !default;
$color: green;

body {
  background: $color;
}
```

... and...

``` scss
$color: blue;
$color: green;

body {
  background: $color;
}
```

**Comment by p2media** [comment]
@jslegers This example will output the same.

But the following one...

``` scss
$color: green;
$color: blue !default;

body {
    background: $color;
}
```

... will generate this CSS:

``` css
body {
    background: green;
}
```

**Comment by koenpunt** [comment]
Not much, but the `!default` is actually used for reverse order.

``` scss
// library.scss
$body-color: blue !default;
```

``` scss
// _vars.scss
$body-color: red;
```

``` scss
// custom.scss
@import "vars";
@import "library";

body {
  color: $body-color;
}
```

**Output:**

``` scss
body {
  color: red;
}
```

So even though `library` is imported later than `vars`, the `!default` flag makes sure that if the variable is already defined, it will not override.

**Comment by jslegers** [comment]
So, the expected behavior for ...

``` scss
$color: blue;
$color: green !default;

main {
  background: $color;
}
```

... is ...

``` css
main {
  background: blue;
}
```

... and  ...

``` scss
$color: blue !default;
$color: green;

main {
  background: $color;
}
```

... is expected to produce ...

``` css
main {
  background: green;
}
```

Right?!

And in the first example, the value for `$color` is `blue` instead of `green` because `$color` already had a value when `$color: green !default;` was processed, right?

**Comment by koenpunt** [comment]
Exactly, but we're getting off topic here :hankey:

**Comment by dehuszar** [comment]
I wrote a PostCSS plugin (and am working on the next set of features) which emulates LESS's `@import (reference)` set of functionality.  I chose to use my own @reference atRule so as to not overload the @import or @extend plugins, which are written by different authors, and would require too much cross-communication between plugins, but if it's helpful to anyone here, [here it is](github.com/dehuszar/postcss-reference).

**Comment by mscarchilli** [comment]
+1
For allowing variable names in import.

This would help cleanup the main components partial when working with some frameworks that have a ton of components.

``` scss
@import "#{$rootpath}/somepath/partial";
```

**Comment by chriseppstein** [comment]
@mscarchilli Dynamic imports are complicated because it makes it impossible to statically analyze the Sass files to determine dependencies, and this means we cannot quickly invalidate compilation results. See #739 for a write up on what is required to support that use case (note: it's complicated)

**Comment by mgcrea** [comment]
:+1: for `import (reference)`, that's really a key feature when you work with large frameworks.

**Comment by JD-Robbs** [comment]
+1 for `import (reference)`.

Just wondering if `@import (reference)` is planned? I can't see any other ways of achieving partial code re-use from other frameworks _unless_ they offer mixins for _everything_.  LESS's `@import (reference)` makes this super simple.

**Comment by veke** [comment]
This would be nice. LESS reference import is really handy.

**Comment by Kazzaky** [comment]
sass has similar functionality like less `import (reference)` and its partial css with leading underscore as first character in name

**Comment by JD-Robbs** [comment]
@Kazzaky while it won't be compiled if you grab a directory containing it, it'll still be compiled into the file importing it, though. Unless there's an undocumented option?

**Comment by Kazzaky** [comment]
@JD-Robbs you are right... my bad

**Comment by JD-Robbs** [comment]
No probs. Just thought that, if I missed it, that'd be a dream come true.

**Comment by grigory-leonenko** [comment]
+1 for import (reference).

At moment I use css modules with sass and foundation, and it critically import only mixins and variables. Because each small module contain all foundation styles.

**Comment by TSMMark** [comment]
+1 `@import (reference) "filename";`

**Comment by xzyfer** [comment]
@nex3 are conditional imports something that could also be shipped in 3.6 independent of the new module system?

**Comment by nex3** [comment]
@xzyfer I'm not sure what you mean by "conditional imports". In general, though, I don't think it makes much sense to add new import features right before doing a big overhaul.

Also, I don't think "3.6" has much meaning at this point, given that Ruby Sass is deprecated.

**Comment by nex3** [comment]
I've updated this issue to track work on [the module system proposal](https://github.com/sass/language/blob/master/proposal/module-system.md). I've unlocked it again so that people can discuss the proposal, but please note that any "+1" comments will be marked as spam.

**Comment by jslegers** [comment]
> @xzyfer I'm not sure what you mean by "conditional imports". In general, though, I don't think it makes much sense to add new import features right before doing a big overhaul.

What he means, is explained by the original issue :

> ### Conditional imports #451
> ```scss
> @if true {
>     @import "foo";
> }
> ```

It's a rather essential feature for anyone who wants to create a configurable CSS framework.

**Comment by glebm** [comment]
For the `@forward "some-module" show/hide thing` syntax, please consider using `only` / `except` instead of `show` / `hide`. Also perhaps `@export` instead of `@forward`.

**Comment by nex3** [comment]
@jslegers I was specifically asking whether Michael was referring to a specific form of conditional imports.

@glebm I wish you'd brought this up when the proposal was first published... I'm planning on declaring it accepted within the next few days, and I don't know if we have time to have a big discussion about the cosmetics before then. It's not impossible to update after it's accepted, though, so can you file an issue on the language repo? Please include your reasoning about why you like your proposed names.

**Comment by jslegers** [comment]
> @jslegers I was specifically asking whether Michael was referring to a specific form of conditional imports.

I can't speak for @xzyfer, but I'd be happy with ANY way to conditionally add imports.

And it seems to me, the format described in the original issue is the "natural" way to do it :

````scss
@if $bool {
    @import "foo";
} @else {
    @import "far";
}
````

**Comment by nex3** [comment]
This is not the place for that discussion, which has been done to death elsewhere.

**Comment by xzyfer** [comment]
Yes my comment was in reference to the issue referenced a few comments
prior. I brought it up at the time because I was looking for something
quick to implement in LibSass.

**Comment by SamanthaAdrichem** [comment]
@use seems to do what @import-once should do? Am I correct in this?

**Comment by MrMcGibblets** [comment]
Trying to follow the issues to get an understanding where this is, in regards to libSass, it was tagged libsass4.0 and is now tagged 5.0, considering 4.0 has a release goal for 2021, are we not expecting this until after then?

**Comment by nex3** [comment]
Closing this out because [LibSass is now deprecated](https://sass-lang.com/blog/libsass-is-deprecated) and we aren't expecting to add any additional features to it.
